Bottom: e5177b117ae926274cc85d940beb90c4b6daea8a
Top:    e893364931a3c926b2eb4297880aa7da4b468558
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2012-11-24 23:02:09 +0000

Play with locked object rendering

(HACK HACK HACK HACK HACK)

Did I mention this was a hacky pile of rubbish code.. the main point of
the exercise is to determine whether the usability is better or not.


---

diff --git a/src/action.c b/src/action.c
index 81e7c1f..a70e4dd 100644
--- a/src/action.c
+++ b/src/action.c
@@ -1131,7 +1131,6 @@ NotifyMode (void)
 	    DrawElement (element);
 	    Draw ();
 	    SetChangedFlag (true);
-	    hid_actionl ("Report", "Object", NULL);
 	  }
 	else if (type != NO_TYPE)
 	  {
@@ -1146,7 +1145,6 @@ NotifyMode (void)
 		Draw ();
 	      }
 	    SetChangedFlag (true);
-	    hid_actionl ("Report", "Object", NULL);
 	  }
 	break;
       }
diff --git a/src/draw.c b/src/draw.c
index 4b7079a..4cb2b2e 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -179,12 +179,15 @@ DrawStrippedText (hidGC gc, ElementType *Element, int min_width)
   gui->graphics->draw_pcb_text (gc, text_ptr, min_width);
 }
 
+void ghid_set_lock_effects (hidGC gc, AnyObjectType *object);
+
 static void
 draw_element_name (ElementType *element)
 {
   if ((TEST_FLAG (HIDENAMESFLAG, PCB) && gui->gui) ||
       TEST_FLAG (HIDENAMEFLAG, element))
     return;
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)element);
   if (doing_pinout || doing_assy)
     gui->graphics->set_color (Output.fgGC, PCB->ElementColor);
   else if (TEST_FLAG (SELECTEDFLAG, &ELEMENT_TEXT (PCB, element)))
@@ -265,6 +268,7 @@ rat_callback (const BoxType * b, void *cl)
 static void
 draw_element_package (ElementType *element)
 {
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)element);
   /* set color and draw lines, arcs, text and pins */
   if (doing_pinout || doing_assy)
     gui->graphics->set_color (Output.fgGC, PCB->ElementColor);
@@ -499,6 +503,7 @@ DrawEMark (ElementType *e, Coord X, Coord Y, bool invisible)
       mark_size = MIN (mark_size, pad0->Thickness / 2);
     }
 
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)e);
   gui->graphics->set_color (Output.fgGC, invisible ? PCB->InvisibleMarkColor : PCB->ElementColor);
   gui->graphics->set_line_cap (Output.fgGC, Trace_Cap);
   gui->graphics->set_line_width (Output.fgGC, 0);
diff --git a/src/draw_funcs.c b/src/draw_funcs.c
index 31f91df..9d2dc91 100644
--- a/src/draw_funcs.c
+++ b/src/draw_funcs.c
@@ -7,6 +7,8 @@
 #include "draw.h"
 #include "hid_draw.h"
 
+void ghid_set_lock_effects (hidGC gc, AnyObjectType *object);
+
 static void
 _draw_pv (PinType *pv, bool draw_hole)
 {
@@ -189,6 +191,7 @@ line_callback (const BoxType * b, void *cl)
   LayerType *layer = cl;
   LineType *line = (LineType *)b;
 
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)line);
   set_layer_object_color (layer, (AnyObjectType *) line);
   dapi->draw_line (line, NULL, NULL);
   return 1;
@@ -200,6 +203,7 @@ arc_callback (const BoxType * b, void *cl)
   LayerType *layer = cl;
   ArcType *arc = (ArcType *)b;
 
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)arc);
   set_layer_object_color (layer, (AnyObjectType *) arc);
   dapi->draw_arc (arc, NULL, NULL);
   return 1;
@@ -216,6 +220,7 @@ poly_callback (const BoxType * b, void *cl)
   struct poly_info *i = cl;
   PolygonType *polygon = (PolygonType *)b;
 
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)polygon);
   set_layer_object_color (i->layer, (AnyObjectType *) polygon);
   dapi->draw_poly (polygon, i->drawn_area, NULL);
   return 1;
@@ -228,6 +233,7 @@ text_callback (const BoxType * b, void *cl)
   TextType *text = (TextType *)b;
   int min_silk_line;
 
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)text);
   if (TEST_FLAG (SELECTEDFLAG, text))
     gui->graphics->set_color (Output.fgGC, layer->SelectedColor);
   else
@@ -244,6 +250,7 @@ text_callback (const BoxType * b, void *cl)
 static void
 set_pv_inlayer_color (PinType *pv, LayerType *layer, int type)
 {
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)pv);
   if (TEST_FLAG (WARNFLAG, pv))          gui->graphics->set_color (Output.fgGC, PCB->WarnColor);
   else if (TEST_FLAG (SELECTEDFLAG, pv)) gui->graphics->set_color (Output.fgGC, (type == VIA_TYPE) ? PCB->ViaSelectedColor
                                                                                                    : PCB->PinSelectedColor);
@@ -279,6 +286,7 @@ pad_inlayer_callback (const BoxType * b, void *cl)
 
   if (ON_SIDE (pad, side))
     {
+      ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)pad);
       if (TEST_FLAG (WARNFLAG, pad))          gui->graphics->set_color (Output.fgGC, PCB->WarnColor);
       else if (TEST_FLAG (SELECTEDFLAG, pad)) gui->graphics->set_color (Output.fgGC, PCB->PinSelectedColor);
       else if (TEST_FLAG (FOUNDFLAG, pad))    gui->graphics->set_color (Output.fgGC, PCB->ConnectedColor);
@@ -299,6 +307,7 @@ pin_hole_callback (const BoxType * b, void *cl)
       (plated == 1 &&  TEST_FLAG (HOLEFLAG, pin)))
     return 1;
 
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)pin);
   set_object_color ((AnyObjectType *) pin, PCB->WarnColor,
                     PCB->PinSelectedColor, NULL, Settings.BlackColor);
 
@@ -316,6 +325,7 @@ via_hole_callback (const BoxType * b, void *cl)
       (plated == 1 &&  TEST_FLAG (HOLEFLAG, via)))
     return 1;
 
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)via);
   set_object_color ((AnyObjectType *) via, PCB->WarnColor,
                     PCB->ViaSelectedColor, NULL, Settings.BlackColor);
 
@@ -326,6 +336,7 @@ via_hole_callback (const BoxType * b, void *cl)
 static int
 pin_callback (const BoxType * b, void *cl)
 {
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)b);
   set_object_color ((AnyObjectType *)b,
                     PCB->WarnColor, PCB->PinSelectedColor,
                     PCB->ConnectedColor, PCB->PinColor);
@@ -337,6 +348,7 @@ pin_callback (const BoxType * b, void *cl)
 static int
 via_callback (const BoxType * b, void *cl)
 {
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)b);
   set_object_color ((AnyObjectType *)b,
                     PCB->WarnColor, PCB->ViaSelectedColor,
                     PCB->ConnectedColor, PCB->ViaColor);
@@ -353,6 +365,7 @@ pad_callback (const BoxType * b, void *cl)
 
   if (ON_SIDE (pad, *side))
     {
+      ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)pad);
       set_object_color ((AnyObjectType *)pad, PCB->WarnColor,
                         PCB->PinSelectedColor, PCB->ConnectedColor,
                         FRONT (pad) ? PCB->PinColor : PCB->InvisibleObjectsColor);
diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index c27554d..d4595b3 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -73,6 +73,8 @@ typedef struct render_priv {
   int subcomposite_stencil_bit;
   char *current_colorname;
   double current_alpha_mult;
+  double current_brightness;
+  double current_saturation;
   GTimer *time_since_expose;
 
   /* Feature for leading the user to a particular location */
@@ -92,6 +94,8 @@ typedef struct hid_gc_struct
 
   const char *colorname;
   double alpha_mult;
+  double brightness;
+  double saturation;
   Coord width;
   gint cap, join;
 }
@@ -309,6 +313,8 @@ ghid_make_gc (void)
   rv->me_pointer = &ghid_hid;
   rv->colorname = Settings.BackgroundColor;
   rv->alpha_mult = 1.0;
+  rv->brightness = 1.0;
+  rv->saturation = 1.0;
   return rv;
 }
 
@@ -499,10 +505,13 @@ set_gl_color_for_gc (hidGC gc)
   hidval cval;
   ColorCache *cc;
   double r, g, b, a;
+  double luminance;
 
   if (priv->current_colorname != NULL &&
       strcmp (priv->current_colorname, gc->colorname) == 0 &&
-      priv->current_alpha_mult == gc->alpha_mult)
+      priv->current_alpha_mult == gc->alpha_mult &&
+      priv->current_brightness == gc->brightness &&
+      priv->current_saturation == gc->saturation)
     return;
 
   free (priv->current_colorname);
@@ -517,6 +526,8 @@ set_gl_color_for_gc (hidGC gc)
 
   priv->current_colorname = strdup (gc->colorname);
   priv->current_alpha_mult = gc->alpha_mult;
+  priv->current_brightness = gc->brightness;
+  priv->current_saturation = gc->saturation;
 
   if (gport->colormap == NULL)
     gport->colormap = gtk_widget_get_colormap (gport->top_window);
@@ -578,6 +589,18 @@ set_gl_color_for_gc (hidGC gc)
 #endif
   }
 
+  r *= gc->brightness;
+  g *= gc->brightness;
+  b *= gc->brightness;
+
+  /* B/W Equivalent brightness */
+  luminance = (r + g + b) / 3.0;
+
+  /* Fade between B/W and colour */
+  r = r * gc->saturation + luminance * (1.0 - gc->saturation);
+  g = g * gc->saturation + luminance * (1.0 - gc->saturation);
+  b = b * gc->saturation + luminance * (1.0 - gc->saturation);
+
   hidgl_flush_triangles (&buffer);
   glColor4d (r, g, b, a);
 }
@@ -596,6 +619,20 @@ ghid_set_alpha_mult (hidGC gc, double alpha_mult)
   set_gl_color_for_gc (gc);
 }
 
+static void
+ghid_set_saturation (hidGC gc, double saturation)
+{
+  gc->saturation = saturation;
+  set_gl_color_for_gc (gc);
+}
+
+static void
+ghid_set_brightness (hidGC gc, double brightness)
+{
+  gc->brightness = brightness;
+  set_gl_color_for_gc (gc);
+}
+
 void
 ghid_set_line_cap (hidGC gc, EndCapStyle style)
 {
@@ -1005,6 +1042,25 @@ ghid_screen_update (void)
 {
 }
 
+void
+ghid_set_lock_effects (hidGC gc, AnyObjectType *object)
+{
+  /* Only apply effects to locked objects when in "lock" mode */
+  if (Settings.Mode == LOCK_MODE &&
+      TEST_FLAG (LOCKFLAG, object))
+    {
+      ghid_set_alpha_mult (gc, 0.8);
+      ghid_set_saturation (gc, 0.3);
+      ghid_set_brightness (gc, 0.7);
+    }
+  else
+    {
+      ghid_set_alpha_mult (gc, 1.0);
+      ghid_set_saturation (gc, 1.0);
+      ghid_set_brightness (gc, 1.0);
+    }
+}
+
 static int
 EMark_callback (const BoxType * b, void *cl)
 {
@@ -1048,6 +1104,7 @@ SetPVColor (PinType *Pin, int Type)
 	color = PCB->PinColor;
     }
 
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)Pin);
   gui->graphics->set_color (Output.fgGC, color);
 }
 
@@ -1075,6 +1132,7 @@ SetPVColor_inlayer (PinType *Pin, LayerType *Layer, int Type)
         color = Layer->Color;
     }
 
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)Pin);
   gui->graphics->set_color (Output.fgGC, color);
 }
 
@@ -1103,6 +1161,7 @@ _draw_pv_name (PinType *pv)
       box.Y1 = pv->Y - pv->Thickness    / 2 + Settings.PinoutTextOffsetY;
     }
 
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)pv);
   gui->graphics->set_color (Output.fgGC, PCB->PinNameColor);
 
   text.Flags = NoFlags ();
@@ -1215,6 +1274,7 @@ draw_pad_name (PadType *pad)
       box.Y1 += Settings.PinoutTextOffsetY;
     }
 
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)pad);
   gui->graphics->set_color (Output.fgGC, PCB->PinNameColor);
 
   text.Flags = NoFlags ();
@@ -1243,6 +1303,7 @@ _draw_pad (hidGC gc, PadType *pad, bool clear, bool mask)
 static void
 draw_pad (PadType *pad)
 {
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)pad);
   if (TEST_FLAG (WARNFLAG | SELECTEDFLAG | FOUNDFLAG, pad))
    {
      if (TEST_FLAG (WARNFLAG, pad))
@@ -1304,6 +1365,7 @@ hole_callback (const BoxType * b, void *cl)
 
   if (TEST_FLAG (HOLEFLAG, pv))
     {
+    ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)pv);
       if (TEST_FLAG (WARNFLAG, pv))
         gui->graphics->set_color (Output.fgGC, PCB->WarnColor);
       else if (TEST_FLAG (SELECTEDFLAG, pv))
@@ -1323,6 +1385,7 @@ hole_callback (const BoxType * b, void *cl)
 static void
 draw_line (LayerType *layer, LineType *line)
 {
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)line);
   if (TEST_FLAG (SELECTEDFLAG | FOUNDFLAG, line))
     {
       if (TEST_FLAG (SELECTEDFLAG, line))
@@ -1345,6 +1408,7 @@ line_callback (const BoxType * b, void *cl)
 static void
 draw_arc (LayerType *layer, ArcType *arc)
 {
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)arc);
   if (TEST_FLAG (SELECTEDFLAG | FOUNDFLAG, arc))
     {
       if (TEST_FLAG (SELECTEDFLAG, arc))
@@ -1372,6 +1436,7 @@ text_callback (const BoxType * b, void *cl)
   TextType *text = (TextType *)b;
   int min_silk_line;
 
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)text);
   if (TEST_FLAG (SELECTEDFLAG, text))
     gui->graphics->set_color (Output.fgGC, layer->SelectedColor);
   else
@@ -1399,6 +1464,7 @@ DrawPlainPolygon (LayerType *Layer, PolygonType *Polygon, const BoxType *drawn_a
     color = PCB->ConnectedColor;
   else
     color = Layer->Color;
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)Polygon);
   gui->graphics->set_color (Output.fgGC, color);
 
   if (gui->graphics->thindraw_pcb_polygon != NULL &&
@@ -1705,6 +1771,7 @@ draw_hole_cyl (PinType *Pin, struct cyl_info *info, int Type)
   else
     color = "drill";
 
+  ghid_set_lock_effects (Output.fgGC, (AnyObjectType *)Pin);
   gui->graphics->set_color (Output.fgGC, color);
   DrawDrillChannel (Pin->X, Pin->Y, Pin->DrillingHole / 2, info->from_layer, info->to_layer, info->scale);
   return 0;
